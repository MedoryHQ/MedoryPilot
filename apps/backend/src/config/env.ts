import { config } from "dotenv";
import { z } from "zod";

config();

const envSchema = z.object({
  PORT: z.string().default("8080"),
  NODE_ENV: z.enum(["development", "production", "test"]).default("test"),

  DATABASE_URL: z.string().url().default("postgres://localhost/testdb"),

  ADMIN_JWT_ACCESS_SECRET: z.string().default("a".repeat(32)),
  ADMIN_JWT_REFRESH_SECRET: z.string().default("b".repeat(32)),
  JWT_ACCESS_SECRET: z.string().default("c".repeat(32)),
  JWT_REFRESH_SECRET: z.string().default("d".repeat(32)),

  CLIENT_URL: z.string().url().default("http://localhost:3000"),
  ADMIN_URL: z.string().url().default("http://localhost:3001"),
  SERVER_URL: z.string().url().default("http://localhost:8080"),
  RESPONSE_URL: z.string().url().default("http://localhost:8080"),
  COOKIE_DOMAIN: z.string().default("localhost"),

  SMS_MOCK: z.string().default("true"),
  SMS_FAIL_GRACEFULLY: z.string().default("true"),
  SENDER_API_KEY: z.string().default("test_api_key"),

  SALT_ROUNDS: z.string().default("10"),
  ADMIN_PASSWORD: z.string().default("admin1234"),
  EMAIL: z.string().email().default("admin@example.com"),
  ADMIN_FIRST_NAME: z.string().default("Admin"),
  ADMIN_LAST_NAME: z.string().default("User"),

  SENDGRID_API_KEY: z.string().default("sendgrid_test_key"),

  SWAGGER_USERNAME: z.string().default("swagger"),
  SWAGGER_PASSWORD: z.string().default("swagger"),

  LOG_LEVEL: z.string().default("info"),

  IP_HASH_SECRET: z.string().min(32).default("e".repeat(32)),
});

const parsed = envSchema.safeParse(process.env);

if (!parsed.success) {
  console.error(
    "Invalid environment variables:",
    parsed.success ? null : parsed.error.format()
  );
  process.exit(1);
}

export const env = parsed.data;

export function getEnvVariable<Key extends keyof typeof env>(
  key: Key
): (typeof env)[Key] {
  return env[key];
}
