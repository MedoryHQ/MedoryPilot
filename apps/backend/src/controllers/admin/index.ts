export * from "./auth";
export * from "./website";
export * from "./upload";
