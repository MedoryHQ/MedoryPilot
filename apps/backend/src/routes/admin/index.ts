export * from "./auth";
export * from "./swagger";
