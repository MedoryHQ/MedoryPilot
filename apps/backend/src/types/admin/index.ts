export * from "./auth";
export * from "./language";
export * from "./website";
