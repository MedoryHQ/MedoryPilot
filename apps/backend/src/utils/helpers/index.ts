export * from "./createTranslations";
export * from "./filterHelper";
export * from "./generateRoutes";
export * from "./isUuid";
export * from "./prismaFilter";
export * from "./utils";
export * from "./validationHelper";
export * from "./responseHelper";
export * from "./email";
