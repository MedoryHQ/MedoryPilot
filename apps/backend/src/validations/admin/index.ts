export * from "./auth.validations";
export * from "./website";
