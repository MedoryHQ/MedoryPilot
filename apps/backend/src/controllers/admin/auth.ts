import { prisma } from "@/config";
import {
  IAdminLogin,
  IForgotAdminPasswordVerification,
  IOtpCodeResend,
  IOtpVerification,
  IResetAdminPassword,
} from "@/types/admin/auth";
import { IForgetPasswordWithEmail } from "@/types/customer";
import {
  generateAccessToken,
  generateRefreshToken,
  sendError,
  verifyField,
  cookieOptions,
  logAdminError as logCatchyError,
  logAdminInfo as logInfo,
  logAdminWarn as logWarn,
  inMinutes,
  getResponseMessage,
  generateSmsCode,
  createPassword,
  generateStageToken,
} from "@/utils";
import { NextFunction, Response, Request } from "express";

export const login = async (
  req: Request,
  res: Response,
  next: NextFunction
) => {
  try {
    const { email, password, remember } = req.body as IAdminLogin;

    logInfo("Login attempt", {
      ip: (req as any).hashedIp,
      path: req.path,
      event: "admin_login_attempt",
    });

    const user = await prisma.admin.findUnique({ where: { email } });

    if (!user || !(await verifyField(password, user.passwordHash))) {
      logWarn("Login failed: invalid credentials", {
        ip: (req as any).hashedIp,
        userId: user?.id,
        event: "admin_login_failed",
      });
      return sendError(req, res, 401, "invalidCredentials");
    }

    if (!user.twoFactorAuth) {
      const payload = { id: user.id, email: user.email };
      const access = generateAccessToken(payload, "ADMIN");
      const refresh = generateRefreshToken(payload, "ADMIN");

      res.cookie("accessToken", access.token, {
        ...cookieOptions,
        maxAge: remember ? refresh.expiresIn : undefined,
      });
      res.cookie("refreshToken", refresh.token, {
        ...cookieOptions,
        maxAge: remember ? refresh.expiresIn : undefined,
      });

      logInfo("Admin logged in successfully", {
        ip: (req as any).hashedIp,
        userId: user.id,
        event: "admin_verified",
      });

      return res.json({
        message: getResponseMessage("loginSuccessful"),
        data: { user: { email: user.email } },
      });
    }

    const {
      hashedSmsCode,
      //  smsCode
    } = await generateSmsCode();

    await prisma.admin.update({
      where: { id: user.id },
      data: {
        smsCode: hashedSmsCode,
        smsCodeExpiresAt: inMinutes(5),
      },
    });

    // await mailer.sendOtpCode(email, smsCode);

    logInfo("Admin login code sent", {
      ip: (req as any).hashedIp,
      userId: user.id,
      event: "admin_login_code_sent",
    });

    const stageToken = generateStageToken(user.id, remember);

    res.cookie("admin_verify_stage", stageToken.token, {
      ...cookieOptions,
      maxAge: 5 * 60 * 1000,
    });

    return res.status(200).json({ message: getResponseMessage("codeSent") });
  } catch (error: unknown) {
    logCatchyError("Login exception", error, {
      ip: (req as any).hashedIp,
      event: "admin_login_exception",
    });
    next(error);
  }
};

export const verifyOtp = async (
  req: Request,
  res: Response,
  next: NextFunction
) => {
  try {
    const { code } = req.body as IOtpVerification;

    const admin = await prisma.admin.findUnique({
      where: { id: (req as any).userId },
    });
    if (!admin) {
      logWarn("Admin verification failed: admin not found", {
        ip: (req as any).hashedIp,
        id: (req as any).userId,
        event: "admin_verification_failed",
      });

      return sendError(req, res, 404, "userNotFound");
    }

    if (
      !admin.smsCodeExpiresAt ||
      new Date(admin.smsCodeExpiresAt) < new Date()
    ) {
      logWarn("Admin verification failed: code expired", {
        ip: (req as any).hashedIp,
        id: (req as any).userId,
        event: "admin_verification_failed",
      });

      return sendError(req, res, 400, "verificationCodeExpired");
    }

    const isValid = admin.smsCode
      ? await verifyField(code, admin.smsCode)
      : false;
    if (!isValid) {
      logWarn("Admin verification failed: code invalid", {
        ip: (req as any).hashedIp,
        id: (req as any).userId,
        event: "admin_verification_failed",
      });

      return sendError(req, res, 401, "smsCodeisInvalid");
    }

    res.clearCookie("admin_verify_stage");

    const remember = (req as any).remember;
    const payload = { id: admin.id, email: admin.email };
    const access = generateAccessToken(payload, "ADMIN");
    const refresh = generateRefreshToken(payload, "ADMIN");

    res.cookie("accessToken", access.token, {
      ...cookieOptions,
      maxAge: remember ? refresh.expiresIn : undefined,
    });

    res.cookie("refreshToken", refresh.token, {
      ...cookieOptions,
      maxAge: remember ? refresh.expiresIn : undefined,
    });

    logInfo("Admin verified successfully", {
      ip: (req as any).hashedIp,
      userId: admin.id,
      event: "admin_verified",
    });

    return res.json({
      message: getResponseMessage("loginSuccessful"),
      data: { user: { email: admin.email } },
    });
  } catch (error) {
    logCatchyError("Admin verify otp exception", error, {
      ip: (req as any).hashedIp,
      userId: req.user?.id,
      event: "admin_verify_otp_exception",
    });
    next(error);
  }
};

export const resendAdminVerificationCode = async (
  req: Request,
  res: Response,
  next: NextFunction
) => {
  try {
    const { email } = req.body as IOtpCodeResend;

    logInfo("Resend verification attempt", {
      ip: (req as any).hashedIp,
      event: "admin_resend_verification_attempt",
    });
    const admin = await prisma.admin.findUnique({
      where: {
        email,
      },
    });

    if (!admin) {
      logWarn("OTP code resend failed: admin not found", {
        ip: (req as any).hashedIp,
        event: "admin_otp_code_resend_failed",
      });
      return sendError(req, res, 404, "userNotFound");
    }
    if (
      admin.smsCodeExpiresAt &&
      new Date(admin.smsCodeExpiresAt) > new Date()
    ) {
      logWarn("OTP code resend failed: code still valid", {
        ip: (req as any).hashedIp,
        event: "admin_otp_code_resend_failed",
      });
      return sendError(req, res, 400, "verificationCodeStillValid");
    }

    const {
      hashedSmsCode,
      //  smsCode
    } = await generateSmsCode();

    await prisma.admin.update({
      where: {
        id: admin.id,
      },
      data: {
        smsCode: hashedSmsCode,
        smsCodeExpiresAt: inMinutes(5),
      },
    });

    // await mailer.sendOtpCode(email, smsCode);

    logInfo("Verification code resent successfully", {
      ip: (req as any).hashedIp,
      userId: admin.id,
      event: "admin_verification_code_resent",
    });

    return res.status(200).json({
      message: getResponseMessage("verificationCodeResent"),
    });
  } catch (error) {
    logCatchyError("Admin Resend verification exception", error, {
      ip: (req as any).hashedIp,
      event: "admin_resend_verification_exception",
    });

    return next(error);
  }
};

export const renew = async (
  req: Request,
  res: Response,
  next: NextFunction
) => {
  try {
    logInfo("Token renew attempt", {
      ip: (req as any).hashedIp,
      userId: req.user?.id,
    });

    const user = await prisma.admin.findUnique({
      where: { id: req.user.id },
      omit: { passwordHash: true },
    });

    if (!user) {
      logWarn("Token renew failed: user not found", {
        ip: (req as any).hashedIp,
        userId: req.user?.id,
      });
      return sendError(req, res, 404, "userNotFound");
    }

    logInfo("Token renew success", {
      ip: (req as any).hashedIp,
      userId: user.id,
    });

    return res.json({
      data: {
        user,
        userType: "ADMIN",
      },
    });
  } catch (error) {
    logInfo("Token renew exception", {
      ip: (req as any).hashedIp,
      userId: req.user?.id,
    });
    next(error);
  }
};

export const forgotPassword = async (
  req: Request,
  res: Response,
  next: NextFunction
) => {
  try {
    const { email } = req.body as IForgetPasswordWithEmail;

    logInfo("Forget password attempt", {
      ip: (req as any).hashedIp,
      event: "admin_forgot_password_attempt",
    });

    const admin = await prisma.admin.findUnique({
      where: {
        email,
      },
    });

    if (!admin) {
      logWarn("Forgot password failed: admin not found", {
        ip: (req as any).hashedIp,
        event: "admin_forgot_password_failed",
      });
      return sendError(req, res, 404, "userNotFound");
    }
    if (
      admin.smsCodeExpiresAt &&
      new Date(admin.smsCodeExpiresAt) > new Date()
    ) {
      logWarn("Forgot password failed: code still valid", {
        ip: (req as any).hashedIp,
        event: "admin_forgot_password_failed",
      });
      return sendError(req, res, 400, "verificationCodeStillValid");
    }

    const {
      hashedSmsCode,
      //  smsCode
    } = await generateSmsCode();

    // await mailer.sendOtpCode(email, smsCode);

    await prisma.admin.update({
      where: {
        id: admin.id,
      },
      data: {
        smsCode: hashedSmsCode,
        smsCodeExpiresAt: inMinutes(5),
      },
    });

    logInfo("Forgot password code sent", {
      ip: (req as any).hashedIp,
      userId: admin.id,
      event: "admin_forgot_password_code_sent",
    });

    return res.status(200).json({
      message: getResponseMessage("codeSent"),
    });
  } catch (error) {
    logCatchyError("Forgot password exception", error, {
      ip: (req as any).hashedIp,
      event: "admin_forgot_password_exception",
    });

    next(error);
  }
};

export const forgotPasswordVerification = async (
  req: Request,
  res: Response,
  next: NextFunction
) => {
  try {
    const { smsCode, email } = req.body as IForgotAdminPasswordVerification;

    logInfo("Forget password verification attempt", {
      ip: (req as any).hashedIp,
      event: "admin_forgot_password_verification_attempt",
    });

    const admin = await prisma.admin.findUnique({
      where: {
        email,
      },
    });

    if (!admin || !admin.smsCode) {
      logWarn("Forgot password verification failed: admin not found", {
        ip: (req as any).hashedIp,
        event: "admin_forgot_password_verification_failed",
      });
      return sendError(req, res, 404, "userNotFound");
    }

    if (
      !admin.smsCodeExpiresAt ||
      new Date(admin.smsCodeExpiresAt) < new Date()
    ) {
      logWarn("Forgot password verification failed: code expired", {
        ip: (req as any).hashedIp,
        event: "admin_forgot_password_verification_failed",
      });
      return sendError(req, res, 400, "verificationCodeExpired");
    }

    const isSmsValid = await verifyField(smsCode, admin.smsCode);
    if (!isSmsValid) {
      logWarn("Forgot password verification failed: invalid code", {
        ip: (req as any).hashedIp,
        event: "admin_forgot_password_verification_failed",
      });
      return sendError(req, res, 401, "smsCodeisInvalid");
    }

    logInfo("Forgot password code verified successfully", {
      ip: (req as any).hashedIp,
      userId: admin.id,
      event: "admin_forgot_password_code_verified",
    });
    return res.status(200).json({
      message: getResponseMessage("codeVerified"),
    });
  } catch (error) {
    logCatchyError("Forgot password verification exception", error, {
      ip: (req as any).hashedIp,
      event: "admin_forgot_password_verification_exception",
    });
    next(error);
  }
};

export const resetPassword = async (
  req: Request,
  res: Response,
  next: NextFunction
) => {
  try {
    const { email, password, smsCode } = req.body as IResetAdminPassword;

    logInfo("Reset password attempt", {
      ip: (req as any).hashedIp,
      event: "admin_reset_password_attempt",
    });

    const admin = await prisma.admin.findUnique({ where: { email } });

    if (!admin || !admin.smsCode) {
      logWarn("Reset password failed: admin not found or no code", {
        ip: (req as any).hashedIp,
        event: "admin_reset_password_failed",
      });
      return sendError(req, res, 404, "userNotFound");
    }

    if (
      !admin.smsCodeExpiresAt ||
      new Date(admin.smsCodeExpiresAt) < new Date()
    ) {
      logWarn("Reset password failed: code expired", {
        ip: (req as any).hashedIp,
        event: "admin_reset_password_failed",
      });
      return sendError(req, res, 400, "verificationCodeExpired");
    }

    const isSmsValid = await verifyField(smsCode, admin.smsCode);
    if (!isSmsValid) {
      logWarn("Reset password failed: invalid code", {
        ip: (req as any).hashedIp,
        event: "admin_reset_password_failed",
      });
      return sendError(req, res, 401, "smsCodeisInvalid");
    }

    const passwordHash = await createPassword(password);

    const newAdmin = await prisma.admin.update({
      where: {
        id: admin.id,
      },
      data: {
        passwordHash,
        smsCode: null,
      },
    });

    logInfo("Password reset successfully", {
      ip: (req as any).hashedIp,
      userId: admin.id,
      event: "admin_password_reset",
    });
    return res.status(200).json({
      message: getResponseMessage("passwordChanged"),
      admin: newAdmin,
    });
  } catch (error) {
    logCatchyError("Reset password exception", error, {
      ip: (req as any).hashedIp,
      event: "admin_reset_password_exception",
    });

    next(error);
  }
};
