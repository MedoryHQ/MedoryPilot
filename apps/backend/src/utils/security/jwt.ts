import jwt from "jsonwebtoken";
import { getEnvVariable, prisma } from "@/config";
import { User } from "@/types/global";

const ACCESS_TOKEN_EXPIRES_IN = "1h";
const REFRESH_TOKEN_EXPIRES_IN = "7d";

const ACCESS_TOKEN_EXPIRES_MS = 60 * 60 * 1000;
const REFRESH_TOKEN_EXPIRES_MS = 7 * 24 * 60 * 60 * 1000;

type TokenType = "USER" | "ADMIN";

interface TokenResponse {
  token: string;
  expiresIn: number;
}

export const generateAccessToken = (
  payload: object,
  type: TokenType
): TokenResponse => {
  const secret =
    type === "USER"
      ? getEnvVariable("JWT_ACCESS_SECRET")
      : getEnvVariable("ADMIN_JWT_ACCESS_SECRET");

  const token = jwt.sign(payload, secret, {
    expiresIn: ACCESS_TOKEN_EXPIRES_IN,
  });

  return {
    token,
    expiresIn: ACCESS_TOKEN_EXPIRES_MS,
  };
};

export const generateRefreshToken = (
  payload: object,
  type: TokenType
): TokenResponse => {
  const secret =
    type === "USER"
      ? getEnvVariable("JWT_REFRESH_SECRET")
      : getEnvVariable("ADMIN_JWT_REFRESH_SECRET");

  const token = jwt.sign(payload, secret, {
    expiresIn: REFRESH_TOKEN_EXPIRES_IN,
  });

  return {
    token,
    expiresIn: REFRESH_TOKEN_EXPIRES_MS,
  };
};

export const verifyRefreshToken = async (refreshToken: string) => {
  const secret = getEnvVariable("JWT_REFRESH_SECRET");

  const storedToken = await prisma.refreshToken.findUnique({
    where: { token: refreshToken },
  });

  if (!storedToken) {
    throw new Error("Refresh token not found or already revoked.");
  }

  if (new Date(storedToken.expiresAt) < new Date()) {
    await prisma.refreshToken.delete({ where: { token: refreshToken } });
    throw new Error("Refresh token has expired.");
  }

  try {
    const decoded = jwt.verify(refreshToken, secret) as {
      id: string;
      phoneNumber: string;
    };
    return decoded;
  } catch (error) {
    throw new Error("Invalid refresh token.");
  }
};

export const checkRefreshToken = (refreshToken: string) => {
  const secret = getEnvVariable("JWT_REFRESH_SECRET");

  try {
    const decoded = jwt.verify(refreshToken, secret) as {
      id: string;
      phoneNumber: string;
    };
    return decoded;
  } catch (error) {
    throw new Error("Invalid or expired refresh token.");
  }
};

export const generateTokens = async (user: User, userType: string) => {
  let accessSecret =
    (userType === "USER"
      ? getEnvVariable("JWT_ACCESS_SECRET")
      : getEnvVariable("ADMIN_JWT_ACCESS_SECRET")) || "";
  let refreshSecret =
    (userType === "USER"
      ? getEnvVariable("JWT_REFRESH_SECRET")
      : getEnvVariable("ADMIN_JWT_REFRESH_SECRET")) || "";

  if (!accessSecret || !refreshSecret) {
    throw new Error("JWT secrets are not provided");
  }

  const accessToken = jwt.sign(user, accessSecret, {
    expiresIn: ACCESS_TOKEN_EXPIRES_IN,
  });

  const refreshToken = jwt.sign(user, refreshSecret, {
    expiresIn: REFRESH_TOKEN_EXPIRES_IN,
  });

  return {
    accessToken,
    refreshToken,
    accessTokenExpires: ACCESS_TOKEN_EXPIRES_MS,
    refreshTokenExpires: REFRESH_TOKEN_EXPIRES_MS,
  };
};
