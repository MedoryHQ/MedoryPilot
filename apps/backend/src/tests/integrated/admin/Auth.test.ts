import request from "supertest";
import express from "express";
import cookieParser from "cookie-parser";
import jwt from "jsonwebtoken";

import { adminAuthRouter } from "@/routes/admin";
import { prisma } from "@/config";
import {
  generateAccessToken,
  generateRefreshToken,
  verifyField,
} from "@/utils";
import { authMatchers } from "@/tests/helpers/authMatchers";
expect.extend(authMatchers);

const app = express();
app.use(express.json());
app.use(cookieParser());
app.use("/admin", adminAuthRouter);

jest.mock("@/config", () => ({
  prisma: {
    admin: {
      findUnique: jest.fn(),
    },
    $disconnect: jest.fn(),
  },
  getEnvVariable: jest.fn((key: string) => {
    if (key === "ADMIN_JWT_ACCESS_SECRET") return "accessSecret";
    if (key === "ADMIN_JWT_REFRESH_SECRET") return "refreshSecret";
    return "test";
  }),
}));

jest.mock("@sendgrid/mail", () => ({
  setApiKey: jest.fn(),
  send: jest.fn(),
}));

jest.mock("@/utils", () => {
  const actual = jest.requireActual("@/utils");

  const errorMsgs = {
    ...((actual as any).errorMessages ?? {}),
    invalidPassword: { en: "Invalid password", ka: "არასწორი პაროლი" },
    passwordLength: { en: "Password length", ka: "პაროლის სიგრძე" },
    invalidEmail: { en: "Invalid email", ka: "ელ-ფოსტა არასწორია" },
    userNotFound: { en: "User not found", ka: "მომხმარებელი ვერ მოიძებნა" },
    noTokenProvided: {
      en: "Access denied. No token provided.",
      ka: "წვდომა უარყოფილია. ტოკენის მოწოდება აუცილებელია",
    },
    invalidCredentials: {
      en: "Invalid email or password",
      ka: "არასწორი ელ-ფოსტა ან პაროლი",
    },
    invalidRefreshToken: {
      en: "Access denied. Invalid refresh token.",
      ka: "წვდომა უარყოფილია. არასწორი refresh ტოკენი.",
    },
    jwtSecretNotProvided: {
      en: "JWT secrets not provided.",
      ka: "JWT secret-ების მოწოდება აუცილებელია.",
    },
  };

  return {
    ...actual,
    getTokenFromRequest: jest.fn((req: any) => {
      const auth = req?.headers?.authorization;
      if (typeof auth === "string" && auth.startsWith("Bearer ")) {
        return auth.split(" ")[1];
      }
      return req?.cookies?.accessToken;
    }),
    selectLogger: jest.fn(() => ({
      warn: jest.fn(),
      info: jest.fn(),
      error: jest.fn(),
    })),
    logAdminInfo: jest.fn(),
    logAdminWarn: jest.fn(),
    logAdminError: jest.fn(),
    getClientIp: jest.fn().mockResolvedValue("hashedIp"),
    verifyField: jest.fn(),
    generateAccessToken: jest.fn(),
    generateRefreshToken: jest.fn(),
    errorMessages: errorMsgs,
  };
});

const mockUser = {
  id: "1",
  email: "admin@test.com",
  passwordHash: "hashedPass",
  name: "Test Admin",
};

afterAll(async () => {
  try {
    await (require("@/config").prisma?.$disconnect?.() ?? Promise.resolve());
  } catch {}
});

describe("Admin auth (integration-style) — /admin/login & /admin/renew", () => {
  beforeEach(() => {
    jest.clearAllMocks();
  });

  describe("POST /admin/login", () => {
    it("logs in with valid credentials and sets cookies", async () => {
      (prisma.admin.findUnique as jest.Mock).mockResolvedValueOnce(mockUser);
      (verifyField as jest.Mock).mockResolvedValueOnce(true);
      (generateAccessToken as jest.Mock).mockReturnValueOnce({
        token: "access123",
        expiresIn: 1000,
      });
      (generateRefreshToken as jest.Mock).mockReturnValueOnce({
        token: "refresh123",
        expiresIn: 2000,
      });

      const res = await request(app)
        .post("/admin/login")
        .send({ email: "admin@test.com", password: "correctPass" });

      expect(res).toHaveStatus(200);
      expect(res.body).toBeValidAdminLoginResponse();
      expect(res.headers["set-cookie"]).toBeDefined();
    });

    it("returns 404 when user not found", async () => {
      (prisma.admin.findUnique as jest.Mock).mockResolvedValueOnce(null);

      const res = await request(app)
        .post("/admin/login")
        .send({ email: "unknown@test.com", password: "Password123" });

      expect(res).toHaveStatus(404);
      expect(res.body.error).toEqual(
        require("@/utils").errorMessages.userNotFound
      );
    });

    it("returns 401 when password invalid", async () => {
      (prisma.admin.findUnique as jest.Mock).mockResolvedValueOnce(mockUser);
      (verifyField as jest.Mock).mockResolvedValueOnce(false);

      const res = await request(app)
        .post("/admin/login")
        .send({ email: "admin@test.com", password: "Password12345" });

      expect(res).toHaveStatus(401);
      expect(res.body.error).toEqual(
        require("@/utils").errorMessages.invalidCredentials
      );
    });

    it("returns 400 when validation fails (missing/invalid fields)", async () => {
      const res = await request(app).post("/admin/login").send({ email: "" });

      expect(res).toHaveStatus(400);
      const params = res.body.errors.map((e: any) => e.param);
      expect(params).toEqual(expect.arrayContaining(["password", "email"]));
    });
    it("matches snapshot on successful login", async () => {
      (prisma.admin.findUnique as jest.Mock).mockResolvedValueOnce(mockUser);
      (verifyField as jest.Mock).mockResolvedValueOnce(true);
      (generateAccessToken as jest.Mock).mockReturnValueOnce({
        token: "access123",
        expiresIn: 1000,
      });
      (generateRefreshToken as jest.Mock).mockReturnValueOnce({
        token: "refresh123",
        expiresIn: 2000,
      });

      const res = await request(app)
        .post("/admin/login")
        .send({ email: "admin@test.com", password: "correctPass" });

      const setCookieHeader = res.headers["set-cookie"];
      const normalizedCookies = Array.isArray(setCookieHeader)
        ? setCookieHeader.map((c: string) =>
            c
              .replace(/Expires=[^;]+/, "Expires=<normalized>")
              .replace(/Max-Age=\d+/, "Max-Age=<normalized>")
              .replace(/accessToken=[^;]+/, "accessToken=<token>")
              .replace(/refreshToken=[^;]+/, "refreshToken=<token>")
          )
        : [];

      const stableBody = {
        ...res.body,
        data: {
          ...res.body.data,
          accessToken: res.body?.data?.accessToken ? "<token>" : undefined,
          refreshToken: res.body?.data?.refreshToken ? "<token>" : undefined,
        },
      };

      expect({
        status: res.status,
        body: stableBody,
        cookies: normalizedCookies,
      }).toMatchSnapshot();
    });
  });

  describe("GET /admin/renew", () => {
    it("returns 401 if no tokens provided", async () => {
      const res = await request(app).get("/admin/renew");

      expect(res).toHaveStatus(401);
      expect(res.body.error).toEqual(
        require("@/utils").errorMessages.noTokenProvided
      );
    });

    it("succeeds with valid access token cookie", async () => {
      const payload = { id: mockUser.id, email: mockUser.email };
      const accessToken = jwt.sign(
        payload,
        (require("@/config").getEnvVariable as jest.Mock)(
          "ADMIN_JWT_ACCESS_SECRET"
        )
      );
      const refreshToken = jwt.sign(
        payload,
        (require("@/config").getEnvVariable as jest.Mock)(
          "ADMIN_JWT_REFRESH_SECRET"
        )
      );

      (prisma.admin.findUnique as jest.Mock).mockResolvedValueOnce({
        ...mockUser,
        passwordHash: undefined,
      });

      const res = await request(app)
        .get("/admin/renew")
        .set("Cookie", [
          `accessToken=${accessToken}`,
          `refreshToken=${refreshToken}`,
        ]);

      expect(res).toHaveStatus(200);
      expect(res.body).toHaveProperty("data");
      expect(res.body.data.user.id).toBe(mockUser.id);
    });

    it("refreshes access token when expired but refresh valid", async () => {
      const jwt = require("jsonwebtoken");
      const payload = { id: "1", email: "admin@test.com" };
      const expiredAccess = jwt.sign(payload, "accessSecret", {
        expiresIn: "-1s",
      });
      const refreshToken = jwt.sign(payload, "refreshSecret");

      (prisma.admin.findUnique as jest.Mock).mockResolvedValueOnce(mockUser);

      const res = await request(app)
        .get("/admin/renew")
        .set("Cookie", [
          `accessToken=${expiredAccess}`,
          `refreshToken=${refreshToken}`,
        ]);

      expect(res.status).toBe(200);
      expect(res.headers["set-cookie"] || []).toEqual(
        expect.arrayContaining([expect.stringContaining("accessToken=")])
      );
    });
    it("returns 401 when refresh token invalid", async () => {
      const payload = { id: mockUser.id, email: mockUser.email };
      const expiredAccess = jwt.sign(
        payload,
        (require("@/config").getEnvVariable as jest.Mock)(
          "ADMIN_JWT_ACCESS_SECRET"
        ),
        {
          expiresIn: "-1s",
        }
      );
      const badRefresh = "not.a.jwt.token";

      const res = await request(app)
        .get("/admin/renew")
        .set("Cookie", [
          `accessToken=${expiredAccess}`,
          `refreshToken=${badRefresh}`,
        ]);

      expect(res).toHaveStatus(401);
      expect(res.body.error).toEqual(
        require("@/utils").errorMessages.invalidRefreshToken
      );
    });

    it("returns 500 when jwt secrets missing (simulated)", async () => {
      const cfg = require("@/config");
      const original = (cfg.getEnvVariable as jest.Mock).mockImplementationOnce(
        () => null
      );

      const payload = { id: mockUser.id, email: mockUser.email };
      const accessToken = jwt.sign(payload, "accessSecret");
      const refreshToken = jwt.sign(payload, "refreshSecret");

      const res = await request(app)
        .get("/admin/renew")
        .set("Cookie", [
          `accessToken=${accessToken}`,
          `refreshToken=${refreshToken}`,
        ]);

      expect(res).toHaveStatus(500);
      expect(res.body.error).toEqual(
        require("@/utils").errorMessages.jwtSecretNotProvided
      );
    });

    it("bubbles controller errors as 500 (mock renew to throw)", async () => {
      const authController = require("@/controllers/admin/auth");
      jest.spyOn(authController, "renew").mockImplementationOnce(() => {
        throw new Error("Unexpected failure");
      });

      const payload = { id: mockUser.id, email: mockUser.email };
      const accessToken = jwt.sign(
        payload,
        (require("@/config").getEnvVariable as jest.Mock)(
          "ADMIN_JWT_ACCESS_SECRET"
        )
      );
      const refreshToken = jwt.sign(
        payload,
        (require("@/config").getEnvVariable as jest.Mock)(
          "ADMIN_JWT_REFRESH_SECRET"
        )
      );

      const res = await request(app)
        .get("/admin/renew")
        .set("Cookie", [
          `accessToken=${accessToken}`,
          `refreshToken=${refreshToken}`,
        ]);

      expect(res.status).toBe(500);
    });

    it("matches snapshot on renew with valid tokens", async () => {
      const payload = { id: mockUser.id, email: mockUser.email };
      const accessToken = jwt.sign(
        payload,
        (require("@/config").getEnvVariable as jest.Mock)(
          "ADMIN_JWT_ACCESS_SECRET"
        )
      );
      const refreshToken = jwt.sign(
        payload,
        (require("@/config").getEnvVariable as jest.Mock)(
          "ADMIN_JWT_REFRESH_SECRET"
        )
      );

      (prisma.admin.findUnique as jest.Mock).mockResolvedValueOnce({
        ...mockUser,
        passwordHash: undefined,
      });

      const res = await request(app)
        .get("/admin/renew")
        .set("Cookie", [
          `accessToken=${accessToken}`,
          `refreshToken=${refreshToken}`,
        ]);

      const setCookieHeader = res.headers["set-cookie"];
      const normalizedCookies = Array.isArray(setCookieHeader)
        ? setCookieHeader.map((c: string) =>
            c
              .replace(/Expires=[^;]+/, "Expires=<normalized>")
              .replace(/Max-Age=\d+/, "Max-Age=<normalized>")
              .replace(/accessToken=[^;]+/, "accessToken=<token>")
              .replace(/refreshToken=[^;]+/, "refreshToken=<token>")
          )
        : [];

      const stableBody = {
        ...res.body,
        data: {
          ...res.body.data,
          accessToken: res.body?.data?.accessToken ? "<token>" : undefined,
          refreshToken: res.body?.data?.refreshToken ? "<token>" : undefined,
        },
      };

      expect({
        status: res.status,
        body: stableBody,
        cookies: normalizedCookies,
      }).toMatchSnapshot();
    });
  });
});
