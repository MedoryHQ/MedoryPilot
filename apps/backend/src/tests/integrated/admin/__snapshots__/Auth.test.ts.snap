// Jest Snapshot v1, https://jestjs.io/docs/snapshot-testing

exports[`Admin auth (integration-style) — /admin/login & /admin/renew GET /admin/renew matches snapshot on renew with valid tokens 1`] = `
{
  "body": {
    "data": {
      "accessToken": undefined,
      "refreshToken": undefined,
      "user": {
        "email": "admin@test.com",
        "id": "1",
        "name": "Test Admin",
      },
      "userType": "ADMIN",
    },
  },
  "cookies": [],
  "status": 200,
}
`;

exports[`Admin auth (integration-style) — /admin/login & /admin/renew POST /admin/login matches snapshot on successful login 1`] = `
{
  "body": {
    "data": {
      "accessToken": "<token>",
      "refreshToken": "<token>",
      "user": {
        "email": "admin@test.com",
        "id": "1",
        "name": "Test Admin",
      },
      "userType": "ADMIN",
    },
  },
  "cookies": [
    "accessToken=<token>; Max-Age=<normalized>; Path=/; Expires=<normalized>; HttpOnly; SameSite=Lax",
    "refreshToken=<token>; Max-Age=<normalized>; Path=/; Expires=<normalized>; HttpOnly; SameSite=Lax",
  ],
  "status": 200,
}
`;
